### PR TITLE
Go mgt samples

### PR DESCRIPTION
**Pull request by avishalom**
Go samples for the admin user management + custom claims

### CODE REVIEW

File: admin/main.go [hunk 1]
```diff
@@ -184,6 +185,176 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	uid := "something"
+	// [START create_user_with_uid]
+	u, err := client.CreateUser(context.Background(),
+		(&auth.UserToCreate{}).UID(uid).Email("user@example.com").PhoneNumber("+15555550100"))
+	// alternatively
+	params := auth.UserToCreate{}
+	// or: 	var params auth.UserToCreate
+	u2, err := client.CreateUser(context.Background(),
+		params.UID(uid).Email("user@example.com").PhoneNumber("+15555550100"))
+	// alternatively
```
Comment (samtstern) [inline admin/main.go]: Is this second `alternatively` comment misplaced?

File: admin/main.go [hunk 2]
```diff
@@ -184,6 +185,176 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	u, err := client.UpdateUser(context.Background(), uid,
+		(&auth.UserToUpdate{}).
+			Email("user@example.com").
+			EmailVerified(true).
+			PhoneNumber("+15555550100").
+			Password("newPassword").
+			DisplayName("John Doe").
+			PhotoURL("http://www.example.com/12345678/photo.png").
+			Disabled(true))
+	// see "CreateUser, above for syntax alteratives to (&auth.UserTo...{})"
```
Comment (samtstern) [inline admin/main.go]: I'd remove this comment since we don't know that this code snippet will always follow the previous one in docs.

**Review by samtstern — APPROVED** [review]
Just some minor comments, but LGTM

File: admin/main.go [hunk 3]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+// https://firebase.google.com/docs/auth/admin/manage-users
+// ==================================================================
+
+func getUser(ctx context.Context, client *auth.Client) *auth.UserRecord {
+	uid := "some_string_uid"
+
+	// [START get_user]
+	u, err := client.GetUser(ctx, uid)
+	if err != nil {
+		log.Fatalf("error getting user %s: %v\n", uid, err)
```
Comment (hiranya911) [inline admin/main.go]: Use Fatalln and Println and remove `\n` literals from the strings.

File: admin/main.go [hunk 4]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	}
+	log.Printf("Successfully fetched user data: %v\n", u)
+	// [END get_user_by_phone]
+	return u
+}
+
+func createUser(ctx context.Context, client *auth.Client) *auth.UserRecord {
+	// [START create_user]
+	u, err := client.CreateUser(context.Background(),
+		(&auth.UserToCreate{}).
```
Comment (hiranya911) [inline admin/main.go]: Please unpack this to multiple lines:

```
r := &auth.UserToCreate{}.
           Email()
           ...
u, err := auth.CreateUser(r)
```

File: admin/main.go [hunk 5]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	log.Printf("Successfully created user: %v\n", u)
+	// [END create_user]
+	return u
+}
+
+func createUserWUID(ctx context.Context, client *auth.Client) *auth.UserRecord {
+	uid := "something"
+	// [START create_user_with_uid]
+	u, err := client.CreateUser(context.Background(),
+		(&auth.UserToCreate{}).UID(uid).Email("user@example.com").PhoneNumber("+15555550100"))
```
Comment (hiranya911) [inline admin/main.go]: Unpack

File: admin/main.go [hunk 6]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	// Alternatively
+	params := auth.UserToCreate{} // also possible: 	var params auth.UserToCreate
+	u2, err := client.CreateUser(context.Background(),
```
Comment (hiranya911) [inline admin/main.go]: Just show this in the snippet. The former style is redundant.

File: admin/main.go [hunk 7]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	log.Printf("Successfully created user: %v\n", u)
+	// [END create_user_with_uid]
+	return u2
+}
+
+func updateUser(ctx context.Context, client *auth.Client) (*auth.UserRecord, *auth.UserRecord) {
+	uid := "d"
+	// [START update_user]
+	u, err := client.UpdateUser(context.Background(), uid,
+		(&auth.UserToUpdate{}).
```
Comment (hiranya911) [inline admin/main.go]: Please unpack. And only show one example. No alternatives necessary.

File: admin/main.go [hunk 8]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+			break
+		}
+		if err != nil {
+			log.Fatalf("error listing users: %s\n", err)
+		}
+		log.Printf("read user user: %v\n", user)
+
+	}
+
+	// Iterating by pages 7 users at a time
```
Comment (hiranya911) [inline admin/main.go]: Change to about 100 instead of 7. That is more realistic.

File: admin/main.go [hunk 9]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+
+	for {
+		user, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			log.Fatalf("error listing users: %s\n", err)
+		}
+		log.Printf("read user user: %v\n", user)
```
Comment (hiranya911) [inline admin/main.go:404]: `log.Println("User: %v", user)`

Follow the log format used in Node.js sample snippets as much as possible.

File: admin/main.go [hunk 10]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	log.Printf("Successfully deleted user: %s\n", uid)
+	// [END delete_user]
+}
+
+func customClaims(ctx context.Context, client *auth.Client) {
+	uid := "uid"
+	// [START custom_claims]
+
+	// erase all existing custom claims
+	err := client.SetCustomUserClaims(context.Background(), uid, nil)
```
Comment (hiranya911) [inline admin/main.go]: Please remove. The snippet here should be equivalent to the Node.js snippet in the same page.

File: admin/main.go [hunk 11]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	if err != nil {
+		log.Fatalf("error removing custom claims %v", err)
+	}
+	// add custom claims
+	err = client.SetCustomUserClaims(context.Background(), uid, map[string]interface{}{"2custom": "2claims"})
```
Comment (hiranya911) [inline admin/main.go]: Use the same claims set in the Node.js example (`admin = true`).

File: admin/main.go [hunk 12]
```diff
@@ -184,6 +185,178 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	if err != nil {
+		log.Fatalf("error setting custom claims %v", err)
+	}
+	// Alternatively
+	_, err = client.UpdateUser(context.Background(), uid,
+		(&auth.UserToUpdate{}).CustomClaims(map[string]interface{}{"2custom": "2claims"}))
+
+	// [END custom_claims]
+
+}
```
Comment (hiranya911) [inline admin/main.go]: Need more sample snippets to shadow the Node.js samples in https://firebase-dot-devsite.googleplex.com/docs/auth/admin/custom-claims

Specifically:
* Accessing claims via decoded ID tokens
* Accessing claims via a call to `GetUser()`
* Defining roles via backend script
* Incrementally updating claims

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Please shadow the Node.js examples as much as possible. The samples in our pages should be functionally equivalent. Also need a few additional snippets to fully demonstrate custom claims support in Go.

File: admin/main.go [hunk 13]
```diff
@@ -184,6 +185,186 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+
+	if err != nil {
+		log.Fatalf("error creating user: %v\n", err)
+	}
+	log.Printf("Successfully created user: %v\n", u)
+	// [END create_user]
+	return u
+}
+
+func createUserWUID(ctx context.Context, client *auth.Client) *auth.UserRecord {
```
Comment (hiranya911) [inline admin/main.go]: Nit: createUserWithUID

File: admin/main.go [hunk 14]
```diff
@@ -184,6 +185,247 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	params := (&auth.UserToCreate{}).
+		Email("user@example.com").
+		EmailVerified(false).
+		PhoneNumber("+15555550100").
+		Password("secretPassword").
+		DisplayName("John Doe").
+		PhotoURL("http://www.example.com/12345678/photo.png").
+		Disabled(false)
+	u, err := client.CreateUser(context.Background(), params)
+
```
Comment (hiranya911) [inline admin/main.go]: Remove whitespace

File: admin/main.go [hunk 15]
```diff
@@ -184,6 +185,247 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+	idToken := "token"
+	// [START verify_custom_claims]
+	// Verify the ID token first.
+	token, err := client.VerifyIDToken(idToken)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	claims := token.Claims
+	if adminVal, found := claims["admin"]; found {
```
Comment (hiranya911) [inline admin/main.go]: Lets simplify this a bit by making a couple of assumptions.

```
if admin, ok := claims["admin"]; ok {
  if admin.(bool) {
    // Allow access ...
  }  
}
```

File: admin/main.go [hunk 16]
```diff
@@ -184,6 +185,247 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+func customClaimsRead(ctx context.Context, client *auth.Client) {
+	uid := "uid"
+	// [START read_custom_user_claims]
+	// Lookup the user associated with the specified uid.
+	user, err := client.GetUser(ctx, uid)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// The claims can be accessed on the user record.
+	if adminVal, found := user.CustomClaims["admin"]; found {
```
Comment (hiranya911) [inline admin/main.go]: Same here

File: admin/main.go [hunk 17]
```diff
@@ -184,6 +185,247 @@ func verifyIDToken(app *firebase.App, idToken string) *auth.Token {
...
+		if err != nil {
+			log.Fatalf("error listing users: %s\n", err)
+		}
+		log.Printf("read user user: %v\n", user)
+	}
+
+	// Iterating by pages 100 users at a time.
+	// Note that using both the Next() function on an iterator and the NextPage()
+	// on a Pager wrapping that same iterator will result in an error.
+	iter2 := client.Users(context.Background(), "")
```
Comment (hiranya911) [inline admin/main.go]: Lets inline the call to client.Users() and remove iter2

**Review by hiranya911 — APPROVED** [review]
LGTM with some nits. Please take a look before merging.

**Pull request merged**